### PR TITLE
Reports improvements

### PR DESCRIPTION
**Pull request by AA-Turner**
As it says on the tin!

Not much functionality change, but a general tidying up so that work can be done more easily in future - reports was the main bit of CI I was stuck on and never finished (and would be most useful to a lot of my other roles).

I co-erced things into the general template we've been using for Hierarchy & Members - a public interface Reports class, and an implementation level ReportsScraper. 

To note - the one real cludge I'm still not massively happy with is in `Reports.__init__` in monkeypatching the `Login._get()` method onto the `ReportsScraper._get()` method - this works, but would like a better solution. This is the second place we've needed the Auth & JK hash stuff - not sure if this will be the last, or if some effort should be made to generalise. (Note generalisation will be difficult I think, as it relies on a lot of stuff in the Login class - maybe a mixin?)

### CODE REVIEW

File: compass/reports.py [hunk 1]
```diff
@@ -1,138 +1,125 @@
...
+    region_disclosure_report = 76
+    region_training_report = 84
+    region_disclosure_management_report = 100
+
+
+class Reports:
+    def __init__(self, session: Logon):
+        """Constructor for Reports."""
+        self._scraper = ReportsScraper(session.s)
+        self._scraper._get = session._get  # massively hacky but we need to send the jk_hash stuff through
```
Comment (AA-Turner) [inline compass/reports.py:37]: xref the last paragraph

**Comment by AA-Turner** [comment]
Note - major TODOS include working timeout extensions for long running exports (default soft t/o is 10min, hard limit is 6hrs, so 35 extensions max), and report location configuration - likely two options, specified level only or level+all down - doesn’t make sense to have many other options here in my opinion.

File: compass/_scrapers/reports.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            with open(filename, "wb") as f:
+                for chunk in r.iter_content(chunk_size=1024 ** 2):  # Chunk size == 1MiB
+                    f.write(chunk)
+
+    def download_report_normal(self, url: str, params: dict, filename: str) -> bytes:
+        start = time.time()
+        csv_export = self._get(url, params=params)
+        print(f"Exporting took {time.time() - start}s")
+        print("Saving report")
+        Path(filename).write_bytes(csv_export.content)  # TODO Debug check
```
Comment (rglss) [inline compass/_scrapers/reports.py:144]: Might be worth checking if this path is valid/writeable

File: compass/reports.py [hunk 3]
```diff
@@ -1,138 +1,124 @@
...
+            (Compass does not include all organisational units in reports by
+            default, and to export all data for a given unit and downwards, we
+            need to add in these missing/unset levels manually).
+        3. We update report configuration data (sent as form data), and check
+            that we are not in an error state.
+        4. We extract the export URL, download the content and save to disk.
+
+        Pitfalls to be aware of in this process include that:
+        - Compass checks user-agent headers in some parts of the process
+            (TODO pinpoint which exactly)
```
Comment (rglss) [inline compass/reports.py:61]: Interesting, is this simply checking for a valid browser UA or that the token request/download is the same UA? As you say, one to explore.

File: compass/utility.py [hunk 4]
```diff
@@ -93,3 +93,36 @@ def parse(date_time_str: str) -> Optional[datetime.datetime]:
...
+        self.thread = threading.Thread(target=self.callback)
+        self.thread.start()
+        return self
+
+    def cancel(self):
+        self.thread.cancel()
+        return self
+
+
+if __name__ == "__main__":
```
Comment (rglss) [inline compass/utility.py:123]: Guessing this is just some test code for when running this class?

**Review by rglss — COMMENTED** [review]
Looks all good, couple of very minor points. Not made use of the reports functionality but from what I've read looks sound 👍

File: compass/_scrapers/reports.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        # In reality we don't care about the output of this POST, just that it doesn't fail
+        report = self._post(run_report, data=form_data, headers={"X-MicrosoftAjax": "Delta=true"})
+        report.raise_for_status()
+
+        # Check error state
+        if "compass.scouts.org.uk%2fError.aspx|" in report.text:
+            raise CompassReportError("Compass Error!")
+
+    def report_keep_alive(self, report_page: str):
+        print(f"Extending Report Session {datetime.datetime.now()}")
```
[the remark this answers is not in the document]
Comment (AA-Turner) [reply compass/_scrapers/reports.py:125]: Yep - legacy & forgot to change

File: compass/utility.py (hunk 4 above)
Comment (AA-Turner) [reply compass/utility.py]: yep, moved to a local script file now

File: compass/reports.py (hunk 3 above)
Comment (AA-Turner) [reply compass/reports.py:61]: It doesn't like requests' default user agent (python-requests/2.25.1, where 2.25.1 is the version) - the javascript evidences user-agent sniffing for front-end stuff so I wouldn't be suprised if it does it still backend - but just doing a stripped down version of the FF user-agent seemed to work. 

Need to track down exactly what breaks though for future ref & potential refactoring!

File: compass/_scrapers/reports.py (hunk 2 above)
Comment (AA-Turner) [reply compass/_scrapers/reports.py:143]: ✅

**Pull request merged**